torch_geometric/data/hetero_data.py

```diff
@@ -616,6 +625,48 @@ def _consistent_size(stores: List[BaseStorage]) -> List[str]:
 
         return data
 
+    # :obj:`FeatureStore` interface ###########################################
+
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Stores a feature tensor in node storage."""
+        if not attr.is_set('index'):
+            attr.index = None
+
+        out = self._node_store_dict.get(attr.group_name, None)
+        if out:
+            # Group name exists, handle index:
+            val = getattr(out, attr.attr_name)
+            if val is not None:
+                val[attr.index] = tensor
+        else:
+            # No group name, just store tensor in node storage:
+            setattr(self[attr.group_name], attr.attr_name, tensor)
+        return True
+
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Obtains a feature tensor from node storage."""
+        # Retrieve tensor and index accordingly:
```

Comment:
In general, should we get `attr.index` first and then `getattr`. The reason I'm asking because it is usually preferred to filter as much data as possible and I think select by index filters more data.

Reply:
Not sure I understand. In this implementation, it doesn't matter since all data is in memory, so we first need to get the tensor and then index that tensor. In an out-of-core implementation, all of `TensorAttr` is passed to the implementor, and they can filter however they like.